### ISSUE TITLE
NTLM authentication on flutter web.

### ISSUE THREAD

**Issue by wterrill** [mrbbot/ntlm#9]
I'd like to request the NTLM package be updated for flutter web, if possible. I've been trying to make it work over the last couple days, and I'd like to know if there was any reason that you didn't try to make this compatible for web. (maybe because you hit some snag?)  I initially thought it was CORS stuff, but after wading through that, I get this errror:

TypeError: Cannot read property 'Symbol(dartx.contains)' of null
    at http.NTLMClient.new._ntlm (http://localhost:50126/packages/ntlm/src/http.dart.lib.js:84:80)
    at _ntlm.next (<anonymous>)
    at http://localhost:50126/dart_sdk.js:37175:33
    at _RootZone.runUnary (http://localhost:50126/dart_sdk.js:37029:58)
    at _FutureListener.thenAwait.handleValue (http://localhost:50126/dart_sdk.js:32116:29)
    at handleValueCallback (http://localhost:50126/dart_sdk.js:32663:49)
    at Function._propagateToListeners (http://localhost:50126/dart_sdk.js:32701:17)
    at _Future.new.[_completeWithValue] (http://localhost:50126/dart_sdk.js:32544:23)
    at async._AsyncCallbackEntry.new.callback (http://localhost:50126/dart_sdk.js:32566:35)
    at Object._microtaskLoop (http://localhost:50126/dart_sdk.js:37290:13)
    at _startMicrotaskLoop (http://localhost:50126/dart_sdk.js:37296:13)
    at http://localhost:50126/dart_sdk.js:32918:9constraints: BoxConstraints(w=885.0, h=821.0)

Any ideas or experience with this?

**Comment by wterrill** [comment]
I did some comparisons between ios and chrome, and found that there seems to be a difference  in the request callback.

In iOS, the  _RequestCallback request in http.dart  in the _ntlm calling parameters) comes up as:
```
Closure: (Map<String, String>) => Future<Response>
call: Closure: (Map<String, String>) => Future<Response>
hashCode: 187108940
runtimeType: Type ((Map<String, String>) => Future<Response>)
```

but in Chrome, it shows up as: 
```
ntlmHeaders => this[_inner].get(url, {headers: ntlmHeaders})
Unknown instance kind: Closure. Please raise a bug against the Dart extension for VS Code.
```
I'm still not sure why, but that seems to be where things start going bad.

**Comment by mrbbot** [comment]
The reason I didn't make this compatible with flutter_web was because at the time I created this, flutter_web didn't exist. 😂 I haven't really tried it out yet but I'll try take a look some time in the next week.

**Comment by wterrill** [comment]
aaaaaa.... duh. that makes a lot more sense. I’d appreciate your help, big time. I’m still poking around the code right now to try and figure out why those two closures look different in the debugger

**Comment by mrbbot** [comment]
Just published a new version (`1.1.0`) of the package with Flutter Web support. If you're making cross-origin requests, you'll need to make sure your server includes the `Access-Control-Expose-Headers: WWW-Authenticate` header in responses. Let me know if you have any issues with this.

**Comment by wterrill** [comment]
Thank you so much for helping, I greatly appreciate it. I went ahead and updated to 1.1.0 and added the WWW-Authenticate to the headers,  but unfortunately, I can't authenticate with it. 

I'm hitting this endpoint: http://12.216.81.220:88/api/AuthenticateUser, with the following NTLM setup: 
   ```
 NTLMClient client = NTLMClient(
      domain: "",
      workstation: "LAPTOP",
      username: "MXOTestAud1",
      password: "Password1",
    ); 

RaisedButton(
                      onPressed: () async {
                        client.get(
                          "http://12.216.81.220:88/api/AuthenticateUser",
                        )
                            .then((res) {
                          print(res.body);
                        }).catchError((Object e) {
                        
                        });
                      },
                      child: Text("Authenticate"),
                    ),
```
Any idea why it is failing?

**Comment by mrbbot** [comment]
I've just tried to run the request locally and I'm getting CORS errors. I think these are probably due to your server not accepting the `Authorization` header containing the challenge. Try adding the `Access-Control-Allow-Headers: Authorization` header to your responses.

**Comment by wterrill** [comment]
I added that header, and it still looks like there is a CORS issue.  

Here are the things I've done to try and resolve it:

1. I added the Access-Control-Allow-Headers:Authorization to the server response.  Here are all the headers I see in the server response:
```
HTTP/1.1 401 Unauthorized
Content-Type: text/html
Server: Microsoft-IIS/8.5
WWW-Authenticate: Negotiate
WWW-Authenticate: NTLM
X-Powered-By: ASP.NET
**Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: POST,OPTIONS,GET
Access-Control-Expose-Headers: WWW-Authenticate
Access-Control-Allow-Headers: Authorization**
Date: Sun, 14 Jun 2020 15:50:13 GMT
Content-Length: 1293
```
2. I also tried using a proxy server by appending: https://cors-anywhere.herokuapp.com/ to my endpoint URL so that it reads:

`client.get("https://cors-anywhere.herokuapp.com/http://12.216.81.220:88/api/AuthenticateUser",)
`
This is definitely an improvement from before... the code isn't crashing when I run it on chrome.  However, it seems like I've added all the CORS headers to the server response that I know about in order to get around this. Are you seeing the same with the above code?
(and by the way, I appreciate the help.  It looks like, at this point, it's a straight-up CORS problem, so any help is definitely above and beyond the norm.)

**Comment by mrbbot** [comment]
Ah I just tried the request in chrome and it's a bit more helpful. Looks like the OPTIONS request that does the CORS checking is not returning an OK status code. On my local test NTLM server, this OPTIONS request returns code 204, not 401.
![image](https://user-images.githubusercontent.com/15955327/84598941-4540e700-ae66-11ea-8d05-ac0d2e8bc77d.png)
![image](https://user-images.githubusercontent.com/15955327/84598950-5853b700-ae66-11ea-9046-07b5376ba39c.png)
This Stack Overflow answer seems relevant: https://stackoverflow.com/a/42859917/4641679

**Comment by wterrill** [comment]
I appreciate you getting back to me, and thanks for the help! I'm still working through CORS issues with my server / browser.  The solution above didn't work, unfortunately.  I'll be sure to post the answer here when I understand what is going on.

**Comment by wterrill** [comment]
So.... wow.  I've learned more about about CORS, and NTLM than I've ever wanted to know.  It looks like the short answer is 'NTLM and CORS don't play very well together.  I had to configure IIS in order to make it work using this stackoverflow post:
 https://stackoverflow.com/questions/50130623/asp-net-core-web-api-using-windows-authentication-cors-request-unauthorised

So, it was a lot more involved than I ever would have imagined.  In any case... thanks for your help!
